[블랙잭 2단계] 동전 미션 제출합니다.

donghyun81 opened the pull request:
# 셀프 체크리스트

<!--
모든 항목을 점검한 후 리뷰를 요청하면, 보다 의미 있는 피드백을 받을 수 있습니다.
-->

- [x] 프로그램이 정상적으로 작동하는가?
- [x] 모든 테스트가 통과하는가?
- [x] 이전에 받은 피드백을 모두 반영하였는가?
- [x] 코딩 스타일 가이드를 준수하였는가?
    - [x] IDE 코드 자동 정렬을 적용하였는가?
    - [x] 린트 검사를 통과하였는가?
- [x] 프로그래밍 요구 사항을 준수하였는가?
- [ ] `README.md`에 기능 목록을 정리하고 명확히 기술하였는가?

# 어떤 부분에 집중하여 리뷰해야 할까요?

Hit에서 딜러와 플레이어의 Stay 전략을 다르게 구현하면서, 코드 중복을 줄일 수 있었습니다.
하지만 아직 익숙하게 사용하는 건 아니라서 필요한 부분 리뷰 해주시면 감사하겠습니다!

## 코드 리뷰 커뮤니케이션

<!--
리뷰어에게 질문이 있다면, PR에 직접 코멘트를 남겨 보세요.  
특정 코드 줄에 의견을 남기면 보다 구체적인 피드백을 받을 수 있습니다.  
-->

📌 [GitHub에서 PR에 댓글 남기는 방법](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request)

## 참고 자료

<!-- 관련 문서나 참고할 만한 자료를 추가하세요. -->

## 스크린샷

<!-- 필요하다면 스크린샷을 첨부하세요. -->

![]()

donghyun81 committed:
docs: 추가 구현 사항 정리 및 피드백 사항 정리

donghyun81 committed:
docs: 배팅 금액 비즈니스 로직 내용 수정

donghyun81 committed:
feat(Money):돈은 마이너스가 되지 않는다.

donghyun81 committed:
feat(BetAmount): 배팅 금액은 판정 결과에 따라 금액을 계산한다.

donghyun81 committed:
feat: 수입 금액으로 흐름 변경

donghyun81 committed:
refactor:상태를 관리해서 결과를 반환하도록 변경

donghyun81 committed:
refactor: 수익 계산 비즈니스 로직으로 이전

donghyun81 committed:
refactor: 테스트 케이스 수정

donghyun81 committed:
refactor: 전체적으로 구조 수정...

laco-dev commented on src/main/kotlin/blackjack/domain/model/MatchResult.kt:
```diff
@@ -0,0 +1,20 @@
+    LOSE(-1.0),
+    DRAW(0.0),
+    BLACKJACK(1.5),
+    BLACKJACK_LOSE(-1.5),
```
호앵... 딜러가 블랙잭이면 오히려 내가 돈을 더 뱉어야하나요?? 😱

donghyun81 replied:
BLACKJACK_LOSE(-1.5)가 추가된 이유는 딜러도 각 플레이어와 수익을 계산할 수 있도록 MatchDealer가 matchResults를 가지고 구현하려고 추가했습니다.
하지만 구현하다 보니 matchPlayers에서 각 bettingPlayer들을 찾아서 다시 matchResults와 비교를 해야 하니 너무 복잡해져서 제거했는데, 이 부분을 제거를 안 했네요.. ㅎㅎ
하고 싶었던 구현은 when문을 제거해서 계산하고
각 플레이어의 대결을 뒤집어서 딜러가 보관하고
딜러의 대결 결과를 통해 딜러의 수익을 정하고 싶었는데 반대로만 남겨놔서 이상한 구현이 되었네요 ㅎㅎ..

laco-dev commented on src/main/kotlin/blackjack/domain/model/betting/BettingPlayer.kt:
```diff
@@ -0,0 +1,17 @@
+
+class BettingPlayer(val name: String, private val money: Money) {
+    fun calculate(matchResult: MatchResult): Profit {
+        return when (matchResult) {
+            MatchResult.WIN -> Profit(money * matchResult.profitRate)
+            MatchResult.LOSE -> Profit(money * matchResult.profitRate)
+            MatchResult.DRAW -> Profit(money * matchResult.profitRate)
+            MatchResult.BLACKJACK -> Profit(money * matchResult.profitRate)
+            MatchResult.BLACKJACK_LOSE -> Profit(money * matchResult.profitRate)
+        }
```
이 로직에 when이 어떤 의미가 있나요?

donghyun81 replied:
승패 결과에 profitRate를 추가하고 나서 when으로 처리했던 부분을 없애야 하는데 생각없이 값만 처리했네요.. ㅎㅎ

laco-dev commented on src/main/kotlin/blackjack/domain/model/InitialParticipant/InitialParticipant.kt:
```diff
@@ -0,0 +1,12 @@
+import blackjack.domain.model.Card
+import blackjack.domain.model.Hands
+
+abstract class InitialParticipant {
+    abstract val name: String
+    protected abstract var hands: Hands
+
+    abstract fun showInitCards(): List<Card>
+
+    fun showCards(count: Int = hands.size) = hands.extractCards(count)
+}
```
상태를 클래스로 나누려는 시도가 좋아 보입니다.

다만, 상태마다 딜러 플레이어 참가자가 매번 쌍으로 만들어져 너무 많은 클래스가 필요합니다.
이외에 상태가 또 생길때마다 N개씩 반복해서 만들어지게 되는데 플레이어든 딜러가 매번 바뀌는게 맞는걸까요?

그렇지 않다면,
딜러와 플레이어가 가지고있는 "카드패"의 상태가 바뀌는 것으로 생각해보면 어떨까요?

딜러
- 카드패 : Initial -> Playing -> Finished

플레이어
- 카드패 : Initial -> Playing -> Finished

donghyun81 replied:
1단계 미션에서 다형성에서 느낀 장점은 ＂추상화된 타입에서 구현체들이 다른 일을 처리 할 수 있다＂ 를 제일 큰 장점으로 느꼈습니다.
하지만 수업에서 상태에 따라 처리하는 부분에서 조건을 줄이는 방법을 보면서 다형성에 장점에 관해서 다르게 느끼는 부분도 많고,
오히려 다형성이 뭔지 더 모호해진 느낌을 받았습니다.
이번에 클래스로 상태를 다 나누어 보고 페로로 피드백도 받아보면서, 왜 카드를 상태에 따라 처리했는지,
카드와 플레이어 중 상태에 따라 나누는 게 누구여야 하는지 등등 조금씩 감을 잡아가는 것 같아서 그에 따라 구현해보겠습니다.

laco-dev commented on src/main/kotlin/blackjack/domain/model/playing/PlayingDealer.kt:
```diff
@@ -0,0 +1,29 @@
+class PlayingDealer(override var hands: Hands, override val name: String = DEALER_NAME) : PlayingParticipant() {
+    constructor(vararg card: Card) : this(Hands(card.toList()))
+
+    override fun getHandsState(): HandState {
+        val score = getScore()
+        val handState =
+            when {
+                score == BUST_THRESHOLD && isStartCardCount() -> HandState.BLACKJACK
+                score > BUST_THRESHOLD -> HandState.BUST
+                score > DEALER_DRAW_THRESHOLD -> HandState.STAY
+                else -> HandState.HIT
+            }
+        return handState
+    }
```
이러한 HandState는 플레잉중이 아니더라도 항상 만들어낼 수 있는 개념이라고 생각해요.
Initial이어도 HandState를 계산할 수 있고 Match여도 계산할 수 있습니다.

laco-dev commented on src/main/kotlin/blackjack/domain/model/service/InitService.kt:
```diff
@@ -0,0 +1,19 @@
+import blackjack.domain.model.Hands
+import blackjack.domain.model.Hands.Companion.START_CARD_COUNT
+
+class InitService(private val playersName: Set<String>, private val deck: Deck) {
+    fun initPlayingParticipants(): InitialParticipants {
+        val initialPlayers =
+            playersName.map { name ->
+                InitialPlayer(Hands(List(START_CARD_COUNT) { deck.draw() }), name)
+            }
+        val playingDealer = InitialDealer(Hands(List(START_CARD_COUNT) { deck.draw() }))
+        return InitialParticipants(playingDealer, initialPlayers)
+    }
+}
```
서비스를 두개로 쪼개면 어떤 이점이 있나요..?
게임의 준비 단계도 하나의 게임 시나리오중 하나인데 플레잉 참가자로 바꾸는건 또 개별적으로 하고 있고 Init 서비스가 무슨 일을 하기 위한 객체인지 잘 이해가 안되요

donghyun81 replied:
각 관심사를 가진 플레이어 끼리 서비스를 처리하면, 누가 게임시나리오 흐름에서 뭘 하는지 명확히 할 수 있고 수정 사항이 있어도 각 서비스에서만 수정 할 수 있지 않을까 싶어서 나눠 봤습니다.
init이랑 play를 하는 플레이어가 같은 관심사를 가지고 있어서 의미는 없어보이지만 ㅎㅎ..
역할 분리를 통한 재사용성에 관해서 솔직히 실질적으로 프로젝트를 만들고 확장하거나 수정해본적이 없어서 이론을 가지고 분리해야하는데 경험이 없으니 이론을 정립하거나 공감하기 어려워서 이것저것 해보는것이 좋다고 생각해서 해봤습니다!

laco-dev commented on src/test/kotlin/blackjack/domain/model/BetTest.kt:
```diff
@@ -0,0 +1,28 @@
+package blackjack.domain.model
+
+class BetTest {
+//    private lateinit var bet: Bet
+//
+//    @BeforeEach
+//    fun setUp() {
+//        bet = Bet(PlayingPlayer("동전"), Money(1000.0))
+//    }
+//
+//    @Test
+//    fun `승리 수익 금액을 계산한다`() {
+//        val actual = 1000.0
+//        assertThat(bet.calculate(MatchResult.WIN).value).isEqualTo(actual)
+//    }
+//
+//    @Test
+//    fun `패배 수익 금액을 계산한다`() {
+//        val actual = -1000.0
+//        assertThat(bet.calculate(MatchResult.LOSE).value).isEqualTo(actual)
+//    }
+//
+//    @Test
+//    fun `무승부 수익 금액을 계산한다`() {
+//        val actual = 0.0
+//        assertThat(bet.calculate(MatchResult.DRAW).value).isEqualTo(actual)
+//    }
```
모두 완료되지 않았더라도 주석을 그대로 남기고 올리지는 말아주세요 :)

donghyun81 replied:
일이 있어서 피드백을 일요일 점심이후로 적용해볼려고 빨리 올리다보니 급하게 행동한 것 같습니다 ㅎㅎ... 주의하겠습니다

laco-dev commented on src/main/kotlin/blackjack/domain/model/profit/ProfitPlayer.kt:
```diff
@@ -0,0 +1,5 @@
+
+import blackjack.domain.model.Profit
+
+class ProfitPlayer(val name: String, val profit: Profit)
```
게임 결과에 수익률 외에 다른 개념들도 필요하면 클래스 이름을 이어붙일건가요???
```
class MatchResultProfitPlayerWithXXX()
```

donghyun81 replied:
수익을 가지고 있는 클래스 보다는 수익과 관련된 행동을 하는 Player클래스라고 생각하고 명칭했지만, 공감 받기 너무 어려운 작명이였던 것 같습니다.
페로로는 명칭할때 행동에 따른 명칭은 어떤 부분을 신경쓰면서 작성해야 공감이 간다고 생각하시나요?

laco-dev replied:
질문이 잘 이해가 안되네요.

근본적으로 `수익과 관련된 행동을 하는 Player`라는게 왜 있는지 모르겠어요.
플레이어에게 수익금을 계산하라는 요청을 하는게 먼저아닌가여?

donghyun81 replied:
음.. 그러네요 구현할때는 수익과 관련된 행동을 하는 Player라고 생각하고 만들었는데,
정리하다보니 상태만 남았던 Player여서 행동에 의미가 전혀 없는 클래스가 되었군요, 수정하겠습니다!

laco-dev reviewed (changes requested):
> 플레이어를 각 흐름 별로 정리했는데 개선할 부분이나 생각할 부분 리뷰 해주시면 감사하겠습니다.

필요한 동작마다 하나하나 클래스로 만들어주셨는데 이런 구조로 한달정도 지나면 클래스가 수백개 수천개가 나올것처럼 보입니다.

플레이어와 딜러라는 특성은 하나뿐인데 배팅 플레이어 수익 플레이어가 어색하지는 않았나요..?

다형성을 지키면서 기능을 확장한게 아닌 그냥 모든 것을 클래스로 만든것 같습니다.
정말 플레이어와 딜러를 수도 없이 만들어내는게 맞는지 다시 고민해보세요.

donghyun81 commented:
> > 플레이어를 각 흐름 별로 정리했는데 개선할 부분이나 생각할 부분 리뷰 해주시면 감사하겠습니다.
> 
> 필요한 동작마다 하나하나 클래스로 만들어주셨는데 이런 구조로 한달정도 지나면 클래스가 수백개 수천개가 나올것처럼 보입니다.
> 
> 플레이어와 딜러라는 특성은 하나뿐인데 배팅 플레이어 수익 플레이어가 어색하지는 않았나요..?
> 
> 다형성을 지키면서 기능을 확장한게 아닌 그냥 모든 것을 클래스로 만든것 같습니다. 정말 플레이어와 딜러를 수도 없이 만들어내는게 맞는지 다시 고민해보세요.

InitialParticipant랑 playerParticipant를 나누거나 matchParticipant를 나누면서 의문이 많이 들었는데, 인스턴스 변수를 줄이기 위해서 Betting에 관심있는 Player, 수익에 관심있는 Player, 결과에 관심있는 Player를 나누다 보니,
관심있는 상태에 따라 클래스를 나누면 각자 상태에 할 수 있는 역할이 분리되려나? 싶은 생각도 있고,
클래스로 다 만들어 봐야 인터페이스를 만들던 추상 클래스를 만들던 상태에 따라 처리 하는 방법들을 고민해보면서 실질적으로 학습할 수 있을 것 같아서 시도해봤습니다!
페로로 피드백으로 필요한 부분 고민해보면서 수정해보겠습니다.

donghyun81 committed:
refactor:Hands를 State로 관리하도록 변경

donghyun81 committed:
refactor: 테스트 케이스 추가

donghyun81 committed:
refactor: Score를 통해서 점수 관리

donghyun81 committed:
refactor: 핸드 상태를 유저 행동으로 이름 변경

donghyun81 committed:
refactor: 하드코드 의미에 맞는 상수로 추출

donghyun81 committed:
refactor: 필요한 함수만 호출 받을 수 있도록 변경

laco-dev commented on src/main/kotlin/blackjack/domain/model/Score.kt:
```diff
@@ -0,0 +1,26 @@
+
+@JvmInline
+value class Score(val value: Int) {
+    fun isBlackJack(count: Int) = value == MAX_SCORE && count == BLACK_JACK_CARD_COUNT
+
+    fun isBustScore() = value > MAX_SCORE
+
+    fun isMaxScore() = value == MAX_SCORE
+
+    fun isDealerStay() = value > DEALER_MIN_STAY_SCORE
```
버스트나 최대 점수는 이해하겠는데 블랙잭이나 딜러 스테이가 점수랑 무슨 관련이 있나요?

> 블랙잭을 판단하는 것, 딜러가 카드를 더 뽑을 수 있는지 판단하는 것

위 비즈니스 로직을 Score가 책임지는게 적절한가요?

donghyun81 replied:
부적절하다고 생각됩니다! 수정하겠습니다.

laco-dev commented on src/main/kotlin/blackjack/domain/model/betting/BettingPlayers.kt:
```diff
@@ -1,7 +1,5 @@
-    fun findPlayer(name: String): BettingPlayer {
-        return requireNotNull(value.find { player -> player.name == name })
-    }
+    fun findPlayer(name: String): BettingPlayer = requireNotNull(value.find { player -> player.name == name })
```
흔히 find라는 의미는 코틀린에서 찾는다를 뜻해서 대부분 반환 타입이 nullable로만들어져요.
`List.find { ... }`를 찾아보세요 비슷한 함수가 많습니다.

laco-dev commented on src/main/kotlin/blackjack/domain/model/hand/DealerHit.kt:
```diff
@@ -0,0 +1,15 @@
+
+import blackjack.domain.model.Card
+
+class DealerHit(hands: Hands = Hands()) : Playing(hands) {
+    override fun nextState(card: Card): State {
+        hands = hands.nextHand(card)
+        return when {
+            hands.isMaxScore() && hands.isBlackJack(hands.size) -> BlackJack(hands)
+            hands.isBustScore() -> Bust(hands)
+            hands.isDealerStay() -> Stay(hands)
+            else -> this
+        }
+    }
+}
```
Hit에 딜러인지 플레이어인지를 엮는 이점이 있을까요?
만약 다른 Hit를 판단하는 My라는 룰이 생기면 MyHit YouHit DoubleHit 이런거를 계속 만들건가요?

laco-dev commented on src/main/kotlin/blackjack/domain/model/hand/Finished.kt:
```diff
@@ -0,0 +1,13 @@
+
+import blackjack.domain.model.Card
+
+sealed class Finished(hands: Hands) : Initial(hands) {
```
Finished가 Initial을 상속하는게 엄청 모순적이지 않나요?

laco-dev commented on src/main/kotlin/blackjack/domain/model/hand/Finished.kt:
```diff
@@ -0,0 +1,13 @@
+import blackjack.domain.model.Card
+
+sealed class Finished(hands: Hands) : Initial(hands) {
+    override fun nextState(card: Card): State {
+        throw IllegalStateException("종료된 상태에서 카드를 더 추가 할 수 없습니다.")
+    }
```
오류를 발생시키는것도 나쁘지 않은 방법인데 자기 자신을 반환하는 방법도 있습니다.

laco-dev commented on src/main/kotlin/blackjack/domain/model/hand/Hands.kt:
```diff
@@ -0,0 +1,42 @@
+
+    fun isDealerStay() = score().isDealerStay()
+
+    private fun getBonusScore(score: Score): Int {
+        val totalScore = score + BONUS_SCORE
+        if (totalScore.isBustScore() && hasAce()) return BONUS_SCORE
+        return 0
+    }
```
점수로 다 옮겼는데 정작 보너스 점수는 안옮긴...
구현에 일관성이 떨어지는군요

laco-dev commented on src/main/kotlin/blackjack/domain/model/service/PlayingService.kt:
```diff
@@ -1,48 +1,47 @@
         playingParticipant: PlayingParticipant,
-        onHandAction: (PlayingParticipant) -> HandState,
-        onStartStay: (PlayingParticipant) -> Unit,
+        onUserAction: (PlayingParticipant) -> UserChoice,
```
TMI : 일단 무작정 고차함수로 만든 시도는 좋지만
우리는 협력을 위한 가장 기본적인 "인터페이스" 라는 것을 사용 할 수도 있었어요.

donghyun81 replied:
학습 내용에 고차함수에 관련된 내용이 있어서 반영 했는데, 기본적인 인터페이스를 통해서 사용 해보겠습니다!

laco-dev commented on src/main/kotlin/blackjack/domain/model/service/PlayingService.kt:
```diff
@@ -1,48 +1,47 @@
+        val choice = onUserAction(playingParticipant)
+        if (UserChoice.STAY == choice) {
+            playingParticipant.handsState.stay()
+            if (playingParticipant.handsState.cards().size == 2) onPlayerState(playingParticipant)
```
카드가 2장이면 onPlayerState가 호출되는 상관관계가 있나요?
어색하지 않나요?

donghyun81 replied:
첫 초기화 상태일 경우에 호출 되도록 수정하겠습니다!

laco-dev reviewed (changes requested):
인스턴스 변수 3개에 너무 얽매이지 않으셔도 됩니다.
그것보다 클래스 구조가 이상해지는게 오히려 더 큰 문제로 보입니다.

개선 방향성에 대해 어느정도 이해하고 계신 듯 하니 수정한 내용에서 몇가지 의견만 추가로 남겨두었습니다.

다음 미션도 진행해야 하니 적절히 판단해서 진행 해주세요

donghyun81 committed:
refactor: 객체에 책임에 맞춰서 기능 이전

-  Hands에서 점수 구하는 기능 Score로 이전
-  단일 Hit를 사용하고, StayStrategy로 Stay판별 책임 이전
-  블랙잭, 및 Dealer의 Stay판별 기능 Hands로 이전

donghyun81 committed:
refactor: 패턴 별로 패키지 이전

donghyun81 committed:
refactor: BaseState를 통해 기본 상태 정의

- Initial에서 초기에 카드 받고 나서 상태 정의
- isStarted를 통해서 시작 개수에서 사용자가 Stay할 경우 상태를 출력하도록 수정

donghyun81 committed:
refactor: 리필에서 새로운 덱으로 다시 만들도록 수정

donghyun81 committed:
refactor: 덱 로직 수정

donghyun81 committed:
refactor: 고차 함수를 인터페이스로 변경

donghyun81 committed:
test: 타입 변경 실행 테스트

- 타입 캐스트는 가능하지만,추가는 불가능

donghyun81 committed:
refactor: 이름을 통해 값을 얻는다는 의미로 수정

laco-dev reviewed (approved):
고생 많으셨습니다.

지금까지 배운 건 앞으로도 계속 적용되는 개념이니 꾸준히 기억하고 연습해보시면 좋겠어요.
남은 우테코 미션도 화이팅입니다.
